R/decontam.R

```diff
@@ -76,19 +78,65 @@
 #' isContaminant(st, conc=conc, method="frequency", threshold=0.2)
 #' isContaminant(st, conc=conc, neg=neg, method="both", threshold=c(0.1,0.5))
 #'
-isContaminant <- function(seqtab, conc=NULL, neg=NULL,
-                          method=c("auto", "frequency", "prevalence", "combined", "minimum", "either", "both"),
-                          batch=NULL, batch.combine=c("minimum", "product", "fisher"),
-                          threshold = 0.1, normalize=TRUE, detailed=TRUE) {
-  # Validate input
-  if(is(seqtab, "phyloseq")) {
-    ps <- seqtab
-    seqtab <- as(ps@otu_table, "matrix")
-    if(ps@otu_table@taxa_are_rows) { seqtab <- t(seqtab) }
-    if(is.character(conc) && length(conc)==1) { conc <- getFromPS(ps, conc) }
-    if(is.character(neg) && length(neg)==1) { neg <- getFromPS(ps, neg) }
-    if(is.character(batch) && length(batch)==1) { batch <- getFromPS(ps, batch) }
+NULL
+
+#' @rdname isContaminant
+#' @export
+setGeneric("isContaminant", signature = c("seqtab"),
+           function(seqtab, ...)
+             standardGeneric("isContaminant"))
+
+#' @rdname isContaminant
+#' @export
+setMethod("isContaminant", signature = c(seqtab = "ANY"),
+  function(seqtab,
+           conc = NULL,
+           neg = NULL,
+           method = c("auto", "frequency", "prevalence", "combined", "minimum", "either", "both"),
+           batch = NULL,
+           batch.combine = c("minimum", "product", "fisher"),
+           threshold = 0.1,
+           normalize = TRUE,
+           detailed = TRUE){
+    if(is(seqtab, "phyloseq")) {
```

Comment:
If the package is switching to S4 dispatch, would it be more appropriate to have an S4 dispatch method for a phyloseq object?

Reply:
Then you would need to bump `phyloseq` up to `Imports`, since the class definition has to be known for S4 dispatch to work.

This would of course increase the footprint of `decontam` and also all downstream package would depend then on `phyloseq` implicitly. I think you mentioned, that you wanted to avoid this.

Alternatively, you can add the S4 dispatch to `phyloseq`, depending on `decontam`. That would probably a better solution.